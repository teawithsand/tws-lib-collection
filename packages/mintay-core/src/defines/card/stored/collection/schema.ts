import { z } from "zod"

const StoredSdelkaCollectionDataV1Schema = z.object({
	globalId: z.string(),
	title: z.string(),
	description: z.string(),
	createdAtTimestamp: z.number(),
	lastUpdatedAtTimestamp: z.number(),
})

export const StoredSdelkaCollectionDataSchema = z.discriminatedUnion(
	"version",
	[
		z.object({
			version: z.literal(1),
			data: StoredSdelkaCollectionDataV1Schema,
		}),
	],
)

export type StoredSdelkaCollectionData = z.infer<
	typeof StoredSdelkaCollectionDataSchema
>
