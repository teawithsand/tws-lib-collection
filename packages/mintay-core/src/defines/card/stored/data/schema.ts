import { z } from "zod"

const StoredSdelkaCardDataV1Schema = z.object({
	globalId: z.string(),
	content: z.string(),
	createdAtTimestamp: z.number(),
	lastUpdatedAtTimestamp: z.number(),
})

export const StoredSdelkaCardDataSchema = z.discriminatedUnion("version", [
	z.object({
		version: z.literal(1),
		data: StoredSdelkaCardDataV1Schema,
	}),
])

export type StoredSdelkaCardData = z.infer<typeof StoredSdelkaCardDataSchema>
