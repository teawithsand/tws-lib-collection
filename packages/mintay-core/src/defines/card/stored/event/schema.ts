import { z } from "zod"
import { StoredSdelkaCardEventV1Schema } from "./schemaV1"

export const StoredSdelkaCardEventSchema = z.discriminatedUnion("version", [
	z.object({
		version: z.literal(1),
		data: StoredSdelkaCardEventV1Schema,
	}),
])

export type StoredSdelkaCardEvent = z.infer<typeof StoredSdelkaCardEventSchema>
