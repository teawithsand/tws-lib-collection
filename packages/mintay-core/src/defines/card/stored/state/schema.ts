import { z } from "zod"
import { StoredSdelkaCardStateV1Schema } from "./schemaV1"

export const StoredSdelkaCardStateSchema = z.discriminatedUnion("version", [
	z.object({
		version: z.literal(1),
		data: StoredSdelkaCardStateV1Schema,
	}),
])

export type StoredSdelkaCardState = z.infer<typeof StoredSdelkaCardStateSchema>
