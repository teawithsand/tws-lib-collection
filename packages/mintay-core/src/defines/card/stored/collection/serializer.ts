import { SdelkaCollectionData } from "../../sdelkaCollectionData"
import { StoredSdelkaCollectionData } from "./schema"

const neverGuard = (value: never): never => {
	throw new Error(`Unsupported version: ${value}`)
}

export class StoredSdelkaCollectionDataSerializer {
	private constructor() {}

	public static readonly serialize = (
		data: SdelkaCollectionData,
	): StoredSdelkaCollectionData => {
		return {
			version: 1,
			data: {
				globalId: data.globalId,
				title: data.title,
				description: data.description,
				createdAtTimestamp: data.createdAtTimestamp,
				lastUpdatedAtTimestamp: data.lastUpdatedAtTimestamp,
			},
		}
	}

	public static readonly deserialize = (
		storedData: StoredSdelkaCollectionData,
	): SdelkaCollectionData => {
		if (storedData.version === 1) {
			const d = storedData.data
			return {
				globalId: d.globalId,
				title: d.title,
				description: d.description,
				createdAtTimestamp: d.createdAtTimestamp,
				lastUpdatedAtTimestamp: d.lastUpdatedAtTimestamp,
			}
		}
		return neverGuard(storedData.version)
	}
}
