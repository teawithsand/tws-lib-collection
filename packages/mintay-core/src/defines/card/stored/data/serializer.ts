import { SdelkaCardData } from "../../sdelkaCardData"
import { StoredSdelkaCardData } from "./schema"

const neverGuard = (value: never): never => {
	throw new Error(`Unsupported version: ${value}`)
}

export class StoredSdelkaCardDataSerializer {
	private constructor() {}

	public static readonly serialize = (
		data: SdelkaCardData,
	): StoredSdelkaCardData => {
		return {
			version: 1,
			data: {
				globalId: data.globalId,
				content: data.content,
				createdAtTimestamp: data.createdAtTimestamp,
				lastUpdatedAtTimestamp: data.lastUpdatedAtTimestamp,
			},
		}
	}

	public static readonly deserialize = (
		storedData: StoredSdelkaCardData,
	): SdelkaCardData => {
		if (storedData.version === 1) {
			const d = storedData.data
			return {
				globalId: d.globalId,
				content: d.content,
				createdAtTimestamp: d.createdAtTimestamp,
				lastUpdatedAtTimestamp: d.lastUpdatedAtTimestamp,
			}
		}
		return neverGuard(storedData.version)
	}
}
