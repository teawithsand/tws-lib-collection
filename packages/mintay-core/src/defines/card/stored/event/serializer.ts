import { SdelkaAnswer } from "../../answer"
import { SdelkaCardEvent, SdelkaCardEventType } from "../../sdelkaCardEvent"
import { StoredSdelkaCardEvent } from "./schema"
import { StoredSdelkaAnswerV1, StoredSdelkaCardEventTypeV1 } from "./schemaV1"

export class SdelkaCardEventSerializer {
	public static readonly toStoredAnswer = (
		answer: SdelkaAnswer,
	): StoredSdelkaAnswerV1 => {
		switch (answer) {
			case SdelkaAnswer.AGAIN:
				return StoredSdelkaAnswerV1.AGAIN
			case SdelkaAnswer.HARD:
				return StoredSdelkaAnswerV1.HARD
			case SdelkaAnswer.GOOD:
				return StoredSdelkaAnswerV1.GOOD
			case SdelkaAnswer.EASY:
				return StoredSdelkaAnswerV1.EASY
		}
		const _exhaustiveCheck: never = answer
		return _exhaustiveCheck
	}

	public static readonly fromStoredAnswer = (
		answer: StoredSdelkaAnswerV1,
	): SdelkaAnswer => {
		switch (answer) {
			case StoredSdelkaAnswerV1.AGAIN:
				return SdelkaAnswer.AGAIN
			case StoredSdelkaAnswerV1.HARD:
				return SdelkaAnswer.HARD
			case StoredSdelkaAnswerV1.GOOD:
				return SdelkaAnswer.GOOD
			case StoredSdelkaAnswerV1.EASY:
				return SdelkaAnswer.EASY
		}
		const _exhaustiveCheck: never = answer
		return _exhaustiveCheck
	}

	public static readonly serialize = (
		event: SdelkaCardEvent,
	): StoredSdelkaCardEvent => {
		switch (event.type) {
			case SdelkaCardEventType.ANSWER:
				return {
					version: 1,
					data: {
						type: StoredSdelkaCardEventTypeV1.ANSWER,
						answer: SdelkaCardEventSerializer.toStoredAnswer(
							event.answer,
						),
						timestamp: event.timestamp,
					},
				}
		}
		const _exhaustiveCheck: never = event.type
		return _exhaustiveCheck
	}

	public static readonly deserialize = (
		storedEvent: StoredSdelkaCardEvent,
	): SdelkaCardEvent => {
		if (storedEvent.version === 1) {
			const data = storedEvent.data
			switch (data.type) {
				case StoredSdelkaCardEventTypeV1.ANSWER:
					return {
						type: SdelkaCardEventType.ANSWER,
						answer: SdelkaCardEventSerializer.fromStoredAnswer(
							data.answer,
						),
						timestamp: data.timestamp,
					}
			}
			const _exhaustiveCheck: never = data.type
			return _exhaustiveCheck
		}
		const _exhaustiveCheckVersion: never = storedEvent.version
		return _exhaustiveCheckVersion
	}
}
