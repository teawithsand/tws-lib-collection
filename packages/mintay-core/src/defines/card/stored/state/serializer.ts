import { SdelkaCardState } from "../../sdelkaCardState"
import { SdelkaCardQueue } from "../../sdelkaQueue"
import { StoredSdelkaCardState } from "./schema"
import { StoredSdelkaCardQueueV1 } from "./schemaV1"

export class SdelkaCardStateSerializer {
	private constructor() {}

	private static readonly toStoredQueue = (
		queue: SdelkaCardQueue,
	): StoredSdelkaCardQueueV1 => {
		switch (queue) {
			case SdelkaCardQueue.NEW:
				return StoredSdelkaCardQueueV1.NEW
			case SdelkaCardQueue.LEARNING:
				return StoredSdelkaCardQueueV1.LEARNING
			case SdelkaCardQueue.LEARNED:
				return StoredSdelkaCardQueueV1.LEARNED
			case SdelkaCardQueue.RELEARNING:
				return StoredSdelkaCardQueueV1.RELEARNING
		}
		const _exhaustiveCheck: never = queue
		return _exhaustiveCheck
	}

	private static readonly fromStoredQueue = (
		queue: StoredSdelkaCardQueueV1,
	): SdelkaCardQueue => {
		switch (queue) {
			case StoredSdelkaCardQueueV1.NEW:
				return SdelkaCardQueue.NEW
			case StoredSdelkaCardQueueV1.LEARNING:
				return SdelkaCardQueue.LEARNING
			case StoredSdelkaCardQueueV1.LEARNED:
				return SdelkaCardQueue.LEARNED
			case StoredSdelkaCardQueueV1.RELEARNING:
				return SdelkaCardQueue.RELEARNING
		}
		const _exhaustiveCheck: never = queue
		return _exhaustiveCheck
	}

	public static readonly serialize = (
		state: SdelkaCardState,
	): StoredSdelkaCardState => {
		return {
			version: 1,
			data: {
				fsrs: {
					dueTimestamp: state.fsrs.dueTimestamp,
					stability: state.fsrs.stability,
					difficulty: state.fsrs.difficulty,
					elapsedDays: state.fsrs.elapsedDays,
					scheduledDays: state.fsrs.scheduledDays,
					reps: state.fsrs.reps,
					lapses: state.fsrs.lapses,
					state: SdelkaCardStateSerializer.toStoredQueue(
						state.fsrs.state,
					),
					lastReviewTimestamp: state.fsrs.lastReviewTimestamp,
				},
			},
		}
	}

	public static readonly deserialize = (
		stored: StoredSdelkaCardState,
	): SdelkaCardState => {
		if (stored.version === 1) {
			const data = stored.data
			return {
				fsrs: {
					dueTimestamp: data.fsrs.dueTimestamp,
					stability: data.fsrs.stability,
					difficulty: data.fsrs.difficulty,
					elapsedDays: data.fsrs.elapsedDays,
					scheduledDays: data.fsrs.scheduledDays,
					reps: data.fsrs.reps,
					lapses: data.fsrs.lapses,
					state: SdelkaCardStateSerializer.fromStoredQueue(
						data.fsrs.state,
					),
					lastReviewTimestamp: data.fsrs.lastReviewTimestamp,
				},
			}
		}
		const _exhaustiveCheck: never = stored.version
		return _exhaustiveCheck
	}
}
