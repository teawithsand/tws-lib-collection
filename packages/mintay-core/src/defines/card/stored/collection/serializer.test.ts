import { describe, expect, test } from "vitest"
import { SdelkaCollectionData } from "../../sdelkaCollectionData"
import { StoredSdelkaCollectionData } from "./schema"
import { StoredSdelkaCollectionDataSerializer } from "./serializer"

const exampleData: SdelkaCollectionData = {
	globalId: "col123",
	title: "Example Collection",
	description: "A description of the example collection",
	createdAtTimestamp: 1234567890,
	lastUpdatedAtTimestamp: 1234567899,
}

describe("StoredSdelkaCollectionDataSerializer", () => {
	test("serialize returns correct structure with version 1", () => {
		const serialized =
			StoredSdelkaCollectionDataSerializer.serialize(exampleData)
		expect(serialized.version).toBe(1)
		expect(serialized.data).toEqual(exampleData)
		expect(serialized).toMatchSnapshot()
	})

	test("deserialize returns correct SdelkaCollectionData for version 1", () => {
		const storedData: StoredSdelkaCollectionData = {
			version: 1,
			data: exampleData,
		}
		const deserialized =
			StoredSdelkaCollectionDataSerializer.deserialize(storedData)
		expect(deserialized).toEqual(exampleData)
		expect(deserialized).toMatchSnapshot()
	})

	test("serialize then deserialize returns original value", () => {
		const serialized =
			StoredSdelkaCollectionDataSerializer.serialize(exampleData)
		const deserialized =
			StoredSdelkaCollectionDataSerializer.deserialize(serialized)
		expect(deserialized).toEqual(exampleData)
	})

	test("deserialize then serialize returns original value", () => {
		const storedData: StoredSdelkaCollectionData = {
			version: 1,
			data: exampleData,
		}
		const deserialized =
			StoredSdelkaCollectionDataSerializer.deserialize(storedData)
		const serialized =
			StoredSdelkaCollectionDataSerializer.serialize(deserialized)
		expect(serialized).toEqual(storedData)
	})
})
