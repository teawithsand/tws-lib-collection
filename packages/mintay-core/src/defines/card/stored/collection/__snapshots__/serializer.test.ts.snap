// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StoredSdelkaCollectionDataSerializer > deserialize returns correct SdelkaCollectionData for version 1 1`] = `
{
  "createdAtTimestamp": 1234567890,
  "description": "A description of the example collection",
  "globalId": "col123",
  "lastUpdatedAtTimestamp": 1234567899,
  "title": "Example Collection",
}
`;

exports[`StoredSdelkaCollectionDataSerializer > serialize returns correct structure with version 1 1`] = `
{
  "data": {
    "createdAtTimestamp": 1234567890,
    "description": "A description of the example collection",
    "globalId": "col123",
    "lastUpdatedAtTimestamp": 1234567899,
    "title": "Example Collection",
  },
  "version": 1,
}
`;
